### ISSUE TITLE
Component's children URN should be relative to the parent

### ISSUE THREAD

**Issue by Sytten** [pulumi/pulumi#4625]
When creating a component, the children resources still need to have globally unique name. I dot not understand the need for this since it is possible make it unique relative to its parent.
For example this cannot be created:
```
     └─ server                          prod-server
 +      ├─ Secret                       stripe-api-key  create
 +      ├─ Secret                       logdna-api-key  create
 +      │  └─ gcp:secretmanager:Secret  secret                            create     1 error
```
In this case pulumi complains because the resulting URN is:
`urn:pulumi:prod::infrastructure::server$Secret$gcp:secretmanager/secret:Secret::secret`

Is there a way to include the name of the parent in there so they dont need to be globally unique?

**Comment by danfhernandez** [comment]
I agree that there should be a way to not have to enforce this convention but rather make it work by default. That said, It's a bit tricky to design it such that there is still control to name the resource.

Currently, it's best practice to use the name passed into the Component in the names of all resources inside of the Component.  See example here in Typescript: 

``` typescript
import { ComponentResource } from "@pulumi/pulumi";

class MyComponent extends ComponentResource {
    
    constructor(name, args, opts) {
        
        const type = “company-name:cloud:resource”; 
        super(type, name, args, opts);
        
        let bucket = new aws.s3.Bucket(`${name}-bucket`, {}, { parent: this });

        this.registerOutputs({
            bucketDomainName: bucket.bucketDomainName,
        });
    }
}
```

**Comment by Sytten** [comment]
That is what I saw in the examples, but I still think its not great. In terraform, the names are relative to the module you are in and it is not a big problem for consistent naming.

**Comment by akergastel** [comment]
I've ran into the same issues, has anything changed on that front / will change?

**Comment by JarnoRFB** [comment]
This feels very awkward to me. I started evaluating pulumi, because I wanted to reduce boilerplate compared to terraform. But having to pass both the `name` and the `parent` for every resource in a component creates a lot more boilerplate and opportunities for inconsistencies. Compare this to terraform where you just put your component code into a separate directory and everything is correctly namespaced and registered.

**Comment by ghostsquad** [comment]
> That is what I saw in the examples, but I still think its not great. In terraform, the names are relative to the module you are in and it is not a big problem for consistent naming.

The problem with how terraform does this, is that it makes refactoring extremely difficult, since the names are based on where they are located (how you compose your abstractions together). It really hurts refactoring, and requires lots of `terraform state mv`.

Flat is better than nested.

Given that you are in a general purpose programming language, I prefer that pulumi _not_ make naming opinionated. You can write your own logic that can automatically prepend the hierarchical structured naming you desire.

**Comment by KaoruDev** [comment]
I got around this by passing the `name` as the `type`. I.E:

```ts
class VPC extends ComponentResource {
  constructor(name: string, opts: ComponentResourceOptions = {}) {
    super(`VPCComponent/${name}`, name, undefined, opts)
  }
}
```

Not sure if this will cause issues, but it does allow me to avoid prefixing the children with the name of the parent.

https://github.com/pulumi/pulumi/discussions/9250

**Comment by ghostsquad** [comment]
The name of the resource really probably shouldn't be global anyway, since the URN is what actually matters. When you import, you need to provide both.

**Comment by simonkarman** [comment]
> The problem with how terraform does this, is that it makes refactoring extremely difficult, since the names are based on where they are located (how you compose your abstractions together). It really hurts refactoring, and requires lots of `terraform state mv`.

@ghostsquad Actually moving a resource to another parent structure is not straightforward in Pulumi either, since the URN **is** based on the parent structure. The difference is that in Pulumi URN is based on the `types` in the chain of parents and not the `names` of those parents.

IMHO: it would make more sense to include the name in the URN, since that makes it easy to create child resources, since you don't have to think about possible name clashes up the chain.

**Comment by diranged** [comment]
> > The problem with how terraform does this, is that it makes refactoring extremely difficult, since the names are based on where they are located (how you compose your abstractions together). It really hurts refactoring, and requires lots of `terraform state mv`.
> 
> @ghostsquad Actually moving a resource to another parent structure is not straightforward in Pulumi either, since the URN **is** based on the parent structure. The difference is that in Pulumi URN is based on the `types` in the chain of parents and not the `names` of those parents.
> 
> IMHO: it would make more sense to include the name in the URN, since that makes it easy to create child resources, since you don't have to think about possible name clashes up the chain. This is also how AWS CDK solves this problem.

I am really surprised that the URN doesn't include the name ... that completely confused me. Especially coming from a CDK world where this is really the default behavior. This really should be fixed, we should't have to remember to set these dynamically built IDs everywhere.
